### PR TITLE
Convert the raylet client (the code in local_scheduler_client.cc) to proper C++.

### PR DESCRIPTION
**Pull request by suquark**
Since I am fixing the possible problems caused by async APIs (moving notification sockets to raylet), I will change the client API. Before introducing new APIs, it is a good chance to improve the client interface.

## What do these changes do?

Convert the raylet client (the code in local_scheduler_client.cc) to proper C++.

The changes include:

1. Use `raylet` instead of `local_scheduler`. This renames some classes, functions and files.
2. Create `RayletClient` and `RayletConnection` class.
3. Update python & java wrappers
4. Remove legacy code (gpu_ids)
5. Use smart pointers instead of plain pointers when necessary.
6. Uniform function interfaces (use `ray::Status` as return values)

BTW, I want to put Cython interface stuffs into another PR after this is merged.

## Related issue number

#3014

Remaining issues are listed in https://github.com/ray-project/ray/issues/3526.

### CODE REVIEW

File: src/ray/raylet/local_scheduler_client.cc
Comment (pcmoritz) [inline src/ray/raylet/local_scheduler_client.cc]: can we make the socket parameter a const std::string&

File: src/ray/raylet/local_scheduler_client.cc
Comment (pcmoritz) [inline src/ray/raylet/local_scheduler_client.cc]: same here

File: src/ray/raylet/local_scheduler_client.cc
Comment (pcmoritz) [inline src/ray/raylet/local_scheduler_client.cc]: make the socket a const std::string&

File: src/ray/raylet/local_scheduler_client.cc
Comment (pcmoritz) [inline src/ray/raylet/local_scheduler_client.cc]: conn -> conn_ and make it a unique_ptr

File: src/ray/raylet/local_scheduler_client.cc
Comment (pcmoritz) [inline src/ray/raylet/local_scheduler_client.cc]: make conn a unique ptr and remove this

File: src/ray/raylet/local_scheduler_client.h
Comment (pcmoritz) [inline src/ray/raylet/local_scheduler_client.h]: we can probably get rid of this

**Comment by pcmoritz** [comment]
Yeah, let's rename LocalSchedulerClient to RayletClient, and also rename the methods to CamelCase and let each of them return a ray::Status (which could be Status::OK by default) to be consistent with the rest of the codebase.

File: src/ray/raylet/local_scheduler_client.h
Comment (robertnishihara) [inline src/ray/raylet/local_scheduler_client.h]: While we're at it, can we format all of these in the style

```
// Connect to the local scheduler.
//
// \param local_scheduler_socket The name of the socket to use to connect to the
// local scheduler.
// \param worker_id A unique ID to represent the worker.
// \param is_worker Whether this client is a worker. If it is a worker, an
// additional message will be sent to register as one.
// \param driver_id The ID of the driver. This is non-nil if the client is a
// driver.
// \return The connection information.
```

**Comment by robertnishihara** [comment]
cc @guoyuhong

File: src/ray/raylet/lib/python/raylet_extension.cc
Comment (robertnishihara) [inline src/ray/raylet/lib/python/raylet_extension.cc]: @suquark @pcmoritz the Python function names should still be formatted like `submit_task` (lower case with underscores) I think.

File: src/ray/raylet/raylet_client.cc
Comment (robertnishihara) [inline src/ray/raylet/raylet_client.cc]: `/*` -> `//`

Comment (robertnishihara) [reply src/ray/raylet/raylet_client.cc]: In other places as well.

File: src/ray/raylet/raylet_client.h
Comment (robertnishihara) [inline src/ray/raylet/raylet_client.h]: We probably don't want this class. Instead, we should just use the `ClientConnection` class that we already have.

@pcmoritz can you comment?

File: src/ray/raylet/raylet_client.h
Comment (robertnishihara) [inline src/ray/raylet/raylet_client.h]: Should be formatted like 

```
/// \param is_worker Whether this client is a worker. If it is a worker, an
/// additional message will be sent to register as one.
```

Comment (robertnishihara) [reply src/ray/raylet/raylet_client.h]: applies in many places

File: src/ray/raylet/raylet_client.h
Comment (robertnishihara) [inline src/ray/raylet/raylet_client.h]:
```suggestion

  /// Register with raylet.
```

applies in many places

File: src/ray/raylet/raylet_client.h
Comment (robertnishihara) [inline src/ray/raylet/raylet_client.h]: need to include `#include <vector>` at top, right?

File: src/ray/raylet/raylet_client.h
Comment (robertnishihara) [inline src/ray/raylet/raylet_client.h]: should we make `client_id`, `is_worker`, and `language` into `const` fields? @pcmoritz

File: src/ray/raylet/raylet_client.h
Comment (robertnishihara) [inline src/ray/raylet/raylet_client.h]: All fields be private.

File: src/ray/raylet/raylet_client.h
Comment (robertnishihara) [inline src/ray/raylet/raylet_client.h]: Given the ownership model, should we return a `unique_ptr`? cc @pcmoritz

File: src/ray/raylet/raylet_client.h
Comment (robertnishihara) [reply src/ray/raylet/raylet_client.h]: and `unordered_map`

File: src/ray/raylet/raylet_client.cc
Comment (robertnishihara) [reply src/ray/raylet/raylet_client.cc]: actually, maybe this code should just go away anyway and use `ClientConnection` instead.

File: src/ray/raylet/raylet_client.h
Comment (robertnishihara) [inline src/ray/raylet/raylet_client.h]: Is there a reason for this to be a separate method? Why not just move the code to the constructor and get rid of this?

File: src/ray/raylet/raylet_client.h
Comment (suquark) [reply src/ray/raylet/raylet_client.h]: However, here's some issues:
1. We need to keep locks.
2. I do not know how to establish a connection for socket path with `ClientConnection`.

Comment (pcmoritz) [reply src/ray/raylet/raylet_client.h]: Yeah, it should be possible to make these work. Let's look into this together tomorrow :)

File: src/ray/raylet/raylet_client.h
Comment (pcmoritz) [reply src/ray/raylet/raylet_client.h]: agreed

File: src/ray/raylet/lib/python/raylet_extension.cc
Comment (suquark) [reply src/ray/raylet/lib/python/raylet_extension.cc]: ok, I have restored their names.

File: src/ray/raylet/raylet_client.h
Comment (zhijunfu) [inline src/ray/raylet/raylet_client.h]: local scheduler  -> raylet

Comment (zhijunfu) [reply src/ray/raylet/raylet_client.h]: there are also a few other places in this file that still refers to local scheduler

File: src/ray/raylet/raylet_client.h
Comment (zhijunfu) [inline src/ray/raylet/raylet_client.h]: for these helper functions that doesn't involve class member variables, maybe good to have them as non-member-functions and thus don't need to appear in header file?

File: src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc
Comment (zhijunfu) [inline src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc]: replace conn with client? 
there are a few other places in this file

File: src/ray/raylet/raylet_client.h
Comment (eric-jj) [inline src/ray/raylet/raylet_client.h]: const std::vector<int>& GetGPUIDs() const{ return gpu_ids_; }

Comment (eric-jj) [inline src/ray/raylet/raylet_client.h]: const std::vector<int>& GetGPUIDs() const{ return gpu_ids_; }

Comment (raulchen) [inline src/ray/raylet/raylet_client.h]: Since we've removed legacy code, this function can also be removed?

File: src/ray/raylet/raylet_client.h
Comment (raulchen) [inline src/ray/raylet/raylet_client.h]: this function should also be const and return reference.

File: src/ray/raylet/raylet_client.h
Comment (raulchen) [reply src/ray/raylet/raylet_client.h]: Converting to `ClientConnection` is great, but may require large changes. Maybe it's okay to do that in a different PR?

File: src/ray/raylet/raylet_client.cc
Comment (eric-jj) [inline src/ray/raylet/raylet_client.cc]: Since we have changed to use the c++, should we change the function to ray::Status RayletConnection::ReadMessage(MessageType type, std::unique_ptr<uint8_t[]> &message)

File: src/ray/raylet/raylet_client.cc
Comment (raulchen) [inline src/ray/raylet/raylet_client.cc]: nit, add a space before `[`

File: src/ray/raylet/lib/python/raylet_extension.cc
Comment (raulchen) [inline src/ray/raylet/lib/python/raylet_extension.cc]:
```cpp
stream << "[RayletClient] FetchOrReconstruct failed: "
       << "raylet client may be closed, check raylet status. error message: " << status.ToString();
```

File: src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc
Comment (raulchen) [inline src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc]: I think this would fail to compile, because variable `client` is re-defined. maybe rename the second `client` to `raylet_client`.

File: src/ray/raylet/lib/python/raylet_extension.cc
Comment (guoyuhong) [inline src/ray/raylet/lib/python/raylet_extension.cc]: `RAY_CHECK_OK_PREPEND` looks too strong here. If there is a unsuccessful disconnect call, the client don't need to crash.

File: src/ray/raylet/lib/python/raylet_extension.cc
Comment (guoyuhong) [inline src/ray/raylet/lib/python/raylet_extension.cc]: FYI. I have a PR. If this message changes, the test should be updated accordingly.

Comment (guoyuhong) [reply src/ray/raylet/lib/python/raylet_extension.cc]: https://github.com/ray-project/ray/pull/3518

File: src/ray/raylet/lib/python/raylet_extension.cc
Comment (guoyuhong) [reply src/ray/raylet/lib/python/raylet_extension.cc]: It looks like `RAY_CHECK_OK_PREPEND` are used everywhere. It really too strict to use it. However, I think it is fine in this PR. We can change it later like https://github.com/ray-project/ray/pull/3493 . It is better to make the caller decide whether to catch this exception instead of crashing by itself.

File: src/ray/raylet/raylet_client.cc
Comment (guoyuhong) [inline src/ray/raylet/raylet_client.cc]: Here is resource leak. `close(socket_fd);` should be called before returning such as Line 44.

**Review by guoyuhong — COMMENTED** [review]
I have left some comments. Thanks, for your effort.

File: src/ray/raylet/lib/python/raylet_extension.cc
Comment (raulchen) [reply src/ray/raylet/lib/python/raylet_extension.cc]: Since now every method in RayletClient returns a status, we can change all `RAY_CHECK_OK_PREPEND` to another macro that throws an exception upwards if status is not ok, and let the caller decide what to do. 
One motivation of having this ability is that sometimes an actor may run background threads, it doesn't want to crash immediately when raylet connection is closed. 
I can work on this once this PR is merged.

Comment (guoyuhong) [reply src/ray/raylet/lib/python/raylet_extension.cc]: Yes, another PR could be clearer.

**Comment by suquark** [comment]
Jenkins retest this please

**Comment by suquark** [comment]
ready to merge. the jenkins errors seem not related

**Review by raulchen — APPROVED** [review]
LGTM, leave it to @robertnishihara to take another look.

**Comment by pcmoritz** [comment]
Looks like our jenkins frontent is down right now @shaneknapp, I'll merge it once we know the tests look ok.

**Pull request merged**
